### ISSUE TITLE
Provide debug interface

### ISSUE THREAD

**Issue by pimlie** [libdns/libdns#124]
It would be great if libdns could provide a default debug interface / logger.

I am currently experiencing API issues with a dns provider that returns a 400 Bad Request for what ever reason, but atm there seems no way to debug this without recompiling both caddy, the caddy module for this provider and the libdns module for this provider...

The ultimate goal would be that when users specify [debug](https://caddyserver.com/docs/caddyfile/options#debug) in their Caddyfile, it would also log the requests to the dns provider api :)

**Comment by mholt** [comment]
Hmm, I see what you mean. The providers currently are designed to be run without configuration, but I can see why a way to enable a debug mode of sorts would be helpful. We'd probably want a way to pass in a logger.

**Comment by mholt** [comment]
That's interesting. While we probably won't use that (officially), it's basically a `http.RoundTripper` that logs on the `RoundTrip()` method. Which is a very reasonable and simple way to do it.

**Comment by mholt** [comment]
Yeah that's possible... what if we just passed in a `*slog.Logger`?

**Comment by gucci-on-fleek** [comment]
@jandelgado 

> wouldn't passing a logger interface to the provider functions in `context.Context` be an option? That way, the interface would not change. Something like that for example:
> 
> ```go
> // on callers site (Caddy)
> func WithLogger(ctx context.Context, logger libdns.Logger) context.Context { 
>     return context.WithValue(ctx, "logger", logger)
> }
> 
> // in libdns module, provided as a convenience function
> func FromContext(ctx context.Context) libdns.Logger {
>      if logger, ok := ctx.Value("logger").(libdns.Logger); ok {
> 	return logger
>      }
>      return libdns.DefaultLogger()  // dummy implementation
> }
> ```

So if I understand this correctly, this would mainly be a documentation change (“providers should use `ctx.Value("logger")` for any logging purposes”) rather than a code change (no need for `libdns` itself to do anything here, other than maybe add a default value somewhere)? Not that I would be opposed to a code change; it's just that I can't really think of anything that `libdns` itself (rather than a provider) would need to log.

**Comment by jandelgado** [comment]
@mholt : passing in `*slog.Logger` would be absolutely fine.
@gucci-on-fleek: exactly, the interface signatures will not change, it would be mainly a change of documentation (and of course extend the code to pass `*slog.Logger` with the corresponding context)

**Comment by mholt** [comment]
I might move this to v1.1, so we can get 1.0 tagged for the big consumers that are ready and upgrading.
